File: lib/models/quote.dart
import 'dart:convert';

class QuoteItem {
  final int? id;
  final int? quoteId;
  final int? itemId;
  final String itemName;
  final String description;
  final double unitPrice;
  final double quantity; // Changed from int to double
  final double totalPrice;

  QuoteItem({
    this.id,
    this.quoteId,
    this.itemId,
    required this.itemName,
    required this.description,
    required this.unitPrice,
    required this.quantity,
    required this.totalPrice,
  });

  factory QuoteItem.fromMap(Map<String, dynamic> map) {
    return QuoteItem(
      id: map['id'] is String ? int.parse(map['id']) : map['id'],
      quoteId: map['quote_id'] is String ? int.parse(map['quote_id']) : map['quote_id'],
      itemId: map['item_id'] is String ? int.parse(map['item_id']) : map['item_id'],
      itemName: map['item_name'] ?? '',
      description: map['description'] ?? '',
      unitPrice: map['unit_price'] is String ? double.parse(map['unit_price']) : (map['unit_price'] ?? 0.0).toDouble(),
      quantity: map['quantity'] is String ? double.parse(map['quantity']) : (map['quantity'] ?? 1.0).toDouble(),
      totalPrice: map['total_price'] is String ? double.parse(map['total_price']) : (map['total_price'] ?? 0.0).toDouble(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'quote_id': quoteId,
      'item_id': itemId,
      'item_name': itemName,
      'description': description,
      'unit_price': unitPrice,
      'quantity': quantity,
      'total_price': totalPrice,
    };
  }

  Map<String, dynamic> toApiMap() {
    return {
      'item_id': itemId?.toString(),
      'item_name': itemName,
      'description': description,
      'unit_price': unitPrice.toString(),
      'quantity': quantity.toString(),
      'total_price': totalPrice.toString(),
    };
  }

  QuoteItem copyWith({
    int? id,
    int? quoteId,
    int? itemId,
    String? itemName,
    String? description,
    double? unitPrice,
    double? quantity, // Changed from int to double
    double? totalPrice,
  }) {
    return QuoteItem(
      id: id ?? this.id,
      quoteId: quoteId ?? this.quoteId,
      itemId: itemId ?? this.itemId,
      itemName: itemName ?? this.itemName,
      description: description ?? this.description,
      unitPrice: unitPrice ?? this.unitPrice,
      quantity: quantity ?? this.quantity,
      totalPrice: totalPrice ?? this.totalPrice,
    );
  }

  @override
  String toString() {
    return 'QuoteItem{itemName: $itemName, quantity: $quantity, unitPrice: $unitPrice, totalPrice: $totalPrice}';
  }
}



class Quote {
  final int? id;
  final int companyId;
  final int clientId;
  final String quoteNumber;
  final DateTime quoteDate;
  final DateTime? expiryDate;
  final String status; // draft, sent, accepted, rejected, expired
  final double subtotal;
  final double taxAmount;
  final double totalAmount;
  final String? notes;
  final List<QuoteItem> items;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Quote({
    this.id,
    required this.companyId,
    required this.clientId,
    required this.quoteNumber,
    required this.quoteDate,
    this.expiryDate,
    required this.status,
    required this.subtotal,
    required this.taxAmount,
    required this.totalAmount,
    this.notes,
    required this.items,
    this.createdAt,
    this.updatedAt,
  });

  factory Quote.fromMap(Map<String, dynamic> map) {
    
    // Parse quote items if they exist
    List<QuoteItem> quoteItems = [];
    if (map['items'] != null) {
      if (map['items'] is List) {
        quoteItems = (map['items'] as List).map((item) => QuoteItem.fromMap(item)).toList();
      } else if (map['items'] is String) {
        // Handle case where items might be stored as JSON string
        try {
          final itemsJson = json.decode(map['items']);
          if (itemsJson is List) {
            quoteItems = itemsJson.map((item) => QuoteItem.fromMap(item)).toList();
          }
        } catch (e) {
          // Silently handle JSON parsing errors
        }
      }
    }
    
    return Quote(
      id: map['id'] is String ? int.parse(map['id']) : map['id'],
      companyId: map['company_id'] is String ? int.parse(map['company_id']) : map['company_id'],
      clientId: map['client_id'] is String ? int.parse(map['client_id']) : map['client_id'],
      quoteNumber: map['quote_number'] ?? '',
      quoteDate: DateTime.parse(map['quote_date']),
      expiryDate: map['expiry_date'] != null ? DateTime.parse(map['expiry_date']) : null,
      status: map['status'] ?? 'draft',
      subtotal: map['subtotal'] is String ? double.parse(map['subtotal']) : (map['subtotal'] ?? 0.0).toDouble(),
      taxAmount: map['tax_amount'] is String ? double.parse(map['tax_amount']) : (map['tax_amount'] ?? 0.0).toDouble(),
      totalAmount: map['total_amount'] is String ? double.parse(map['total_amount']) : (map['total_amount'] ?? 0.0).toDouble(),
      notes: map['notes'],
      items: quoteItems,
      createdAt: map['created_at'] != null ? DateTime.parse(map['created_at']) : null,
      updatedAt: map['updated_at'] != null ? DateTime.parse(map['updated_at']) : null,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'company_id': companyId,
      'client_id': clientId,
      'quote_number': quoteNumber,
      'quote_date': quoteDate.toIso8601String(),
      'expiry_date': expiryDate?.toIso8601String(),
      'status': status,
      'subtotal': subtotal,
      'tax_amount': taxAmount,
      'total_amount': totalAmount,
      'notes': notes,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  Map<String, dynamic> toApiMap() {
    return {
      'client_id': clientId.toString(),
      'quote_number': quoteNumber,
      'quote_date': quoteDate.toIso8601String().split('T')[0], // YYYY-MM-DD format
      'expiry_date': expiryDate?.toIso8601String().split('T')[0],
      'status': status,
      'subtotal': subtotal.toString(),
      'tax_amount': taxAmount.toString(),
      'total_amount': totalAmount.toString(),
      'notes': notes ?? '',
      'items': items.map((item) => item.toApiMap()).toList(),
    };
  }

  Quote copyWith({
    int? id,
    int? companyId,
    int? clientId,
    String? quoteNumber,
    DateTime? quoteDate,
    DateTime? expiryDate,
    String? status,
    double? subtotal,
    double? taxAmount,
    double? totalAmount,
    String? notes,
    List<QuoteItem>? items,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Quote(
      id: id ?? this.id,
      companyId: companyId ?? this.companyId,
      clientId: clientId ?? this.clientId,
      quoteNumber: quoteNumber ?? this.quoteNumber,
      quoteDate: quoteDate ?? this.quoteDate,
      expiryDate: expiryDate ?? this.expiryDate,
      status: status ?? this.status,
      subtotal: subtotal ?? this.subtotal,
      taxAmount: taxAmount ?? this.taxAmount,
      totalAmount: totalAmount ?? this.totalAmount,
      notes: notes ?? this.notes,
      items: items ?? this.items,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  String toString() {
    return 'Quote{id: $id, quoteNumber: $quoteNumber, status: $status, totalAmount: $totalAmount}';
  }
}

File: backend/classes/Quote.php
<?php

class Quote {
    private $db;
    
    public function __construct($database) {
        $this->db = $database;
    }
    
    public function getAll($companyId) {
        try {
            $query = "SELECT q.*, c.name as client_name 
                     FROM quotes q 
                     LEFT JOIN clients c ON q.client_id = c.id 
                     WHERE q.company_id = ? 
                     ORDER BY q.created_at DESC";
            $stmt = $this->db->prepare($query);
            $stmt->execute([$companyId]);
            return $stmt->fetchAll(PDO::FETCH_ASSOC);
        } catch (Exception $e) {
            error_log("Quote getAll error: " . $e->getMessage());
            return [];
        }
    }
    
    public function getById($id, $companyId) {
        try {
            $query = "SELECT q.*, c.name as client_name 
                     FROM quotes q 
                     LEFT JOIN clients c ON q.client_id = c.id 
                     WHERE q.id = ? AND q.company_id = ?";
            $stmt = $this->db->prepare($query);
            $stmt->execute([$id, $companyId]);
            return $stmt->fetch(PDO::FETCH_ASSOC);
        } catch (Exception $e) {
            error_log("Quote getById error: " . $e->getMessage());
            return null;
        }
    }
    
    public function create($data, $companyId) {
        try {
            $this->db->beginTransaction();
            
            // Calculate totals from items
            $items = $data['items'] ?? [];
            $subtotal = 0;
            foreach ($items as $item) {
                $subtotal += floatval($item['total_price'] ?? 0);
            }
            
            // Insert quote (keep items column for backward compatibility)
            $query = "INSERT INTO quotes (company_id, client_id, quote_number, quote_date, expiry_date, status, subtotal, tax_amount, total_amount, notes, items) 
                     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
            $stmt = $this->db->prepare($query);
            $result = $stmt->execute([
                $companyId,
                $data['client_id'],
                $data['quote_number'],
                $data['quote_date'],
                $data['expiry_date'] ?? null,
                $data['status'] ?? 'draft',
                $data['subtotal'] ?? $subtotal,
                $data['tax_amount'] ?? 0,
                $data['total_amount'] ?? ($subtotal + floatval($data['tax_amount'] ?? 0)),
                $data['notes'] ?? null,
                json_encode($items) // Store items as JSON for backward compatibility
            ]);
            
            if (!$result) {
                throw new Exception("Failed to create quote");
            }
            
            $quoteId = $this->db->lastInsertId();
            
            // Insert quote items in separate table
            if (isset($data['items']) && is_array($data['items'])) {
                foreach ($data['items'] as $item) {
                    $this->createQuoteItem($quoteId, $item);
                }
            }
            
            $this->db->commit();
            
            // Return the created quote
            return $this->getById($quoteId, $companyId);
            
        } catch (Exception $e) {
            $this->db->rollBack();
            error_log("Quote create error: " . $e->getMessage());
            throw $e;
        }
    }
    
    public function update($id, $data, $companyId) {
        try {
            $this->db->beginTransaction();
            
            // Calculate totals from items
            $items = $data['items'] ?? [];
            $subtotal = 0;
            foreach ($items as $item) {
                $subtotal += floatval($item['total_price'] ?? 0);
            }
            
            // Update quote
            $query = "UPDATE quotes SET 
                        client_id = ?, 
                        quote_number = ?, 
                        quote_date = ?, 
                        expiry_date = ?, 
                        status = ?, 
                        subtotal = ?, 
                        tax_amount = ?, 
                        total_amount = ?, 
                        notes = ?,
                        items = ?,
                        updated_at = CURRENT_TIMESTAMP
                      WHERE id = ? AND company_id = ?";
            $stmt = $this->db->prepare($query);
            $result = $stmt->execute([
                $data['client_id'],
                $data['quote_number'],
                $data['quote_date'],
                $data['expiry_date'] ?? null,
                $data['status'] ?? 'draft',
                $data['subtotal'] ?? $subtotal,
                $data['tax_amount'] ?? 0,
                $data['total_amount'] ?? ($subtotal + floatval($data['tax_amount'] ?? 0)),
                $data['notes'] ?? null,
                json_encode($items), // Store items as JSON for backward compatibility
                $id,
                $companyId
            ]);
            
            if (!$result) {
                throw new Exception("Failed to update quote");
            }
            
            // Update quote items if provided
            if (isset($data['items']) && is_array($data['items'])) {
                // Delete existing items
                $this->deleteQuoteItems($id);
                
                // Insert new items
                foreach ($data['items'] as $item) {
                    $this->createQuoteItem($id, $item);
                }
            }
            
            $this->db->commit();
            
            // Return the updated quote
            return $this->getById($id, $companyId);
            
        } catch (Exception $e) {
            $this->db->rollBack();
            error_log("Quote update error: " . $e->getMessage());
            throw $e;
        }
    }
    
    public function delete($id, $companyId) {
        try {
            $query = "DELETE FROM quotes WHERE id = ? AND company_id = ?";
            $stmt = $this->db->prepare($query);
            return $stmt->execute([$id, $companyId]);
        } catch (Exception $e) {
            error_log("Quote delete error: " . $e->getMessage());
            return false;
        }
    }
    
    public function getQuoteItems($quoteId) {
        try {
            $query = "SELECT * FROM quote_items WHERE quote_id = ? ORDER BY id";
            $stmt = $this->db->prepare($query);
            $stmt->execute([$quoteId]);
            return $stmt->fetchAll(PDO::FETCH_ASSOC);
        } catch (Exception $e) {
            error_log("Quote getQuoteItems error: " . $e->getMessage());
            return [];
        }
    }
    
    private function createQuoteItem($quoteId, $itemData) {
        $query = "INSERT INTO quote_items (quote_id, item_id, item_name, description, unit_price, quantity, total_price) 
                 VALUES (?, ?, ?, ?, ?, ?, ?)";
        $stmt = $this->db->prepare($query);
        return $stmt->execute([
            $quoteId,
            $itemData['item_id'] ?? null,
            $itemData['item_name'],
            $itemData['description'] ?? '',
            $itemData['unit_price'],
            $itemData['quantity'] ?? 1,
            $itemData['total_price']
        ]);
    }
    
    private function deleteQuoteItems($quoteId) {
        $query = "DELETE FROM quote_items WHERE quote_id = ?";
        $stmt = $this->db->prepare($query);
        return $stmt->execute([$quoteId]);
    }
    
    public function generateQuoteNumber($companyId) {
        try {
            // Get the highest quote number for this company
            $query = "SELECT quote_number FROM quotes WHERE company_id = ? ORDER BY id DESC LIMIT 1";
            $stmt = $this->db->prepare($query);
            $stmt->execute([$companyId]);
            $lastQuote = $stmt->fetch(PDO::FETCH_ASSOC);
            
            if ($lastQuote && preg_match('/QUO-(\d+)/', $lastQuote['quote_number'], $matches)) {
                $nextNumber = intval($matches[1]) + 1;
            } else {
                $nextNumber = 1;
            }
            
            return 'QUO-' . str_pad($nextNumber, 4, '0', STR_PAD_LEFT);
        } catch (Exception $e) {
            error_log("Quote generateQuoteNumber error: " . $e->getMessage());
            return 'QUO-0001';
        }
    }
}
?>
